### DISCUSSION TITLE
actions and questions accessing object within an angular controller. (Leaflet testing)

### DISCUSSION THREAD

**Discussion by scaddenp** [serenity-js/serenity-js#1640] [How to... ?]
I have an application using Leafletjs for mapping and I am struggling with how to test interactions with the map. Eg a mouse hovering over a particular map position should create a pop up; clicks on map tools will change the map content. These are easily emulated by using setters (for action) and getters (for questions) on an instance of an L.Map object. 

What I am baffled with though, is that the map variable ( map = L.Map(etc) ) is buried within am angular controller. It seems I would have to expose it to global scope to use it within a serentityjs test. I think I must be missing something very obvious here and surely it must be a common enough requirement  for angular developers using serentityjs.

**Comment by jan-molak** [comment]
Hi @scaddenp!

There are several ways you could approach it:
- you could, like you say, expose the map to the global scope, then make Protractor get hold of it via [ExecuteScript](https://github.com/jan-molak/serenity-js/blob/237b0e5fdf4856b81289613f1041289971816ff1/packages/serenity-js/src/serenity-protractor/screenplay/interactions/execute.ts#L6)
- you could wrap the map in an Angular service, then get hold of that service in the acceptance test using a combination of [ExecuteScript](https://github.com/jan-molak/serenity-js/blob/237b0e5fdf4856b81289613f1041289971816ff1/packages/serenity-js/src/serenity-protractor/screenplay/interactions/execute.ts#L6) and a technique similar to the one [described here](https://stackoverflow.com/questions/20619225/accessing-angular-inside-protractor-test)

What version of Angular are you using? If you could put together a minimal example on github it would be easier for myself or other members of the Serenity/JS community to propose an approach.

Best,
Jan

**Comment by scaddenp** [comment]
Thanks for those suggestions. Currently using angularjs but looking at transitioning as I refactor to improve the test environment. Leaflet is currently buried in a directive but not "pure" angular so map variable is exposed to controller. I will look at getting a minimal version together if I cant figure it out from your suggestion.

**Comment by jan-molak** [comment]
Sure, please keep me posted.

**Comment by jan-molak** [comment]
*Note*: Serenity/JS 2.0 will come with improved `ExecuteScript` interactions that could be useful here. [Preview](https://github.com/jan-molak/serenity-js/tree/2.0/packages/protractor/spec/screenplay/interactions/execute-script).

**Comment by scaddenp** [comment]
Looks useful. battling away with need to simulate mouse clicks on specific screen x,y locations. There are cases where cannot use the L.Map methods ( a masking DOM element), so having to do very low level in scripts.
